java/arcs/core/util/StateMachine.kt

```diff
@@ -0,0 +1,209 @@
+ * 5) The new state's [onEntry] handler is invoked.
+ * 6) If an exception occurs, [onError] is invoked.
+ * 7) Any [trigger] events that happen during these sequence are queued and processed as the last
+ * step. [onError] flushes pending events.
+ *
+ * [StateMachine] represents the only legal transitions possible. Triggering events
+ * which are not relevant for the current state results in an exception. Invalid state transition
+ * attempts are not silently ignored.
+ *
+ * Loosely inspired by those used in gaming or protocol frameworks, like Fettle
+ * [http://thehiflyer.github.io/Fettle/] but simplified for Arcs needs, and constructed via
+ * idiomatic Kotlin type-safe builders.
+ *
+ * @param STATE usually an enum or sealed class representing possible states
+ * @param EVENT usually an enum or sealed class representing possible event triggered transitions
+ * @param MACHINE the subclass type that extends [StateMachine]
+ * @property state the current/starting state of the machine.
+ */
+abstract class StateMachine<STATE, EVENT, MACHINE : StateMachine<STATE, EVENT, MACHINE>>(
+    var state: STATE
+) {
+    private lateinit var states: Map<STATE, StateNode<STATE, EVENT, MACHINE>>
+    private val eventQueue = mutableListOf<EVENT>()
+
+    /** True if there are no pending events to process. */
+    val idle = eventQueue.isEmpty()
+
+    internal data class Transition<EVENT, STATE, MACHINE>(
+        val nextState: STATE,
+        val block: suspend (MACHINE).(STATE) -> Unit
+    )
+
+    internal data class StateNode<STATE, EVENT, MACHINE>(
+        val entryHandler: EntryHandler<MACHINE>,
+        val exitHandler: EntryHandler<MACHINE>,
+        val errorHandler: ErrorHandler<MACHINE>,
+        val eventHandlers: Map<EVENT, Transition<EVENT, STATE, MACHINE>>
+    )
+
+    /**
+     * Triggers an event to be queued and processed immediately. This should only be called from
+     * external code, never from state machine handlers.
+     */
+    suspend fun triggerImmediate(event: EVENT) {
+        trigger(event)
+        triggerInternal()
+    }
+
+    /**
+     * Called by state machine handlers to queue an event to be processed after processing the
+     * current event is finished processing.
+     */
+    protected suspend fun trigger(event: EVENT) {
+        eventQueue += event
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun triggerInternal() {
+        if (eventQueue.isEmpty()) {
+            return
+        }
+        val event = eventQueue.removeAt(0)
+        val stateNode = requireNotNull(states[state])
+        val eventHandler = requireNotNull(stateNode.eventHandlers[event]) {
+            "Illegal state transition $event while in state $state"
+        }
+
+        try {
+            stateNode.exitHandler.invoke(this as MACHINE)
+            eventHandler.block(this, eventHandler.nextState)
+            transitionTo(eventHandler.nextState)
+        } catch (e: Exception) {
```

Comment:
Any exception? 

A comment explaining why would be good here, and if it's necessary - maybe `Throwable` is better?